pennylane/beta/queuing/operation.py

```diff
@@ -13,47 +13,154 @@
+    'own'.
+    """
+    self._eigvals_cache = None
+    self.obs = []
+
+    for o in args:
+        if isinstance(o, qml.operation.Tensor):
+            self.obs.extend(o.obs)
+
+        elif isinstance(o, qml.operation.Observable):
+            self.obs.append(o)
+
+        QueuingContext.update_info(o, owner=self)
+
+    QueuingContext.append(self, owns=tuple(args))
+
+
+def tensor_matmul(self, other):
+    """Monkeypatched :meth:`~.Tensor.__matmul__` method, to ensure
+    that tensors created via left matrix multiplication are correctly
+    added to the queue, and their components are annotated.
+    """
+    if isinstance(other, qml.operation.Tensor):
+        self.obs.extend(other.obs)
+
+    elif isinstance(other, qml.operation.Observable):
+        self.obs.append(other)
+
+    owning_info = QueuingContext.get_info(self)["owns"] + (other,)
+
+    # update the annotated queue information
+    QueuingContext.update_info(self, owns=owning_info)
+    QueuingContext.update_info(other, owner=self)
+
+    return self
+
+
+def tensor_rmatmul(self, other):
+    """Monkeypatched :meth:`~.Tensor.__rmatmul__` method, to ensure
+    that tensors created via right matrix multiplication are correctly
+    added to the queue, and their components are annotated.
+    """
+    self.obs[:0] = [other]
+    QueuingContext.update_info(other, owner=self)
+    return self
+
+
+def mock_operations():
+    """Create mock operations, observables and tensors that are monkeypatched
+    to work with the new QueuingContext.
+
+    Creating mocked methods, rather than directly monkeypatching/overwriting the methods,
+    allows us to later remove/undo the monkeypatching once no longer needed.
+
+    Returns:
+        list[MagicMock]: list containing the mocked operations.
+    """
+    # Monkeypatch the 'expand' method of operations directly.
+    # This is required since it does not already exist, and so can't be mocked.
+    qml.operation.Operation.expand = operation_expand
+
+    mocks = []
+
+    # create mock operation methods
+    mocks += [mock.patch.object(qml.operation.Operation, "queue", operation_queue)]
+    mocks += [mock.patch.object(qml.operation.Operation, "inv", operation_inv)]
+
+    # create mock observable methods
+    mocks += [mock.patch.object(qml.operation.Observable, "queue", operation_queue)]
+
+    # create mock tensor methods
+    mocks += [mock.patch.object(qml.operation.Tensor, "__init__", tensor_init)]
+    mocks += [mock.patch.object(qml.operation.Tensor, "__matmul__", tensor_matmul)]
+    mocks += [mock.patch.object(qml.operation.Tensor, "__rmatmul__", tensor_rmatmul)]
+
+    # Mock the operations so that they no longer perform validation
+    # on argument types and domain. This is required to avoid the operations
+    # complaining when unknown types (such as TensorFlow and Torch tensors) are
+    # used as arguments.
+    mocks += [mock.patch.object(qml.operation.Operator, "do_check_domain", False)]
+
+    return mocks
```

Comment:
Yes, is it just the old operations + some queuing monkeypatching? Or does anything change?

Reply:
Within the `Tape` context, this changes the following methods:

* `Operation.queue`: to force operations to queue themselves to the beta queue

* `Operation.inv`: to force inverted operations to annotate the queue with their inversion, rather than changing an internal attribute. This allows the _same_ operation object to be used in different queues with different inverse status (which makes more sense! There have been weird edge cases where inverting an operation in one qnode causes it to be inverted _everwhere_).

* `Operation.expand`: since this doesn't previously exist. This is just a wrapper around `Operation.decomposition`, but it returns a tape rather than a list.

* `Tensor.init`: Currently, tensors don't perform any queuing, so we need to make sure tensors both (a) queue themselves, and (b) annotate the queue declaring any previous observables that they 'own'.

* `Tensor.__matmul__`, `Tensor.__rmatmul__`: we need to also patch these as per above.

Once the tape context is exited, all of these patches are removed.

When the tape is moved to core proper (and the old core removed), these 'patches' should be moved to become officially part of the operations.

Reply:
But happy to discuss 'in person' as well!